lib/dynamo.js

```diff
@@ -27,159 +31,153 @@ exports.client = async () => {
+  return exports.updateResult(id, payload, segments, result);
+};
 
-  // duplicate ids cause errors
-  const uniqueIds = ids.filter((id, idx) => ids.indexOf(id) === idx);
+/**
+ * Update multiple items, limiting concurrency
+ */
+exports.updateAll = async (updates, concurrency = 25) => {
+  const result = { success: [], failures: [] };
+  const client = await exports.client();
 
-  const results = [];
-  const keys = uniqueIds.map(id => [process.env.DDB_TABLE, {id: {S: id}}]);
-  const client = await exports.client()
-  try {
-    for await (const item of new BatchGet(client, keys)) {
-      results.push(item)
+  // spin up N workers, and process the queue of updates
+  async function worker() {
+    let args;
+    while ((args = updates.shift())) {
+      try {
+        result.success.push(await exports.update.apply(this, args.concat([client])));
+      } catch (err) {
+        logger.error(`DDB Error [${process.env.DDB_TABLE}]: ${err}`, { args });
+        result.failures.push(args);
+      }
     }
-  } catch (err) {
-    err.message = `DDB ${process.env.DDB_TABLE} - ${err.message}`;
-    throw err;
   }
 
-  // inflate/reassemble payload, and sort by input order
-  const keyedResults = {};
-  await Promise.all(results.map(async ([, data]) => {
-    const id = data.id.S;
-    const buffer = data.payload.B;
-    keyedResults[id] = await inflate(buffer, id, idField);
-  }));
-  return ids.map(id => keyedResults[id] || null);
+  const threads = Array(concurrency).fill(true);
+  await Promise.all(threads.map(() => worker()));
+
+  return result;
 };
 
 /**
- * Batch write items to dynamodb
+ * Get item (for tests)
  */
-exports.write = async (records, idField = 'id') => {
-  if (!idField) {
-    throw new Error('You must specify an id field');
-  }
+exports.get = async id => {
+  const client = await exports.client();
+  const params = { Key: { id: { S: id } }, TableName: process.env.DDB_TABLE };
+  return client.getItem(params).promise();
+};
+
+/**
+ * Delete item (for tests)
+ */
+exports.delete = async id => {
+  const client = await exports.client();
+  const params = { Key: { id: { S: id } }, TableName: process.env.DDB_TABLE };
+  return client.deleteItem(params).promise();
+};
+
+/**
+ * Params for an updateItem request
+ */
+exports.updateParams = async (id, payload, segments) => {
   if (!process.env.DDB_TABLE) {
     throw new Error('You must set a DDB_TABLE');
   }
-  if (!Array.isArray(records)) {
-    return await exports.write([records], idField);
-  }
-  if (!records.every(r => r[idField])) {
-    throw new Error(`All records must include the field '${idField}'`);
-  }
 
-  // compress payload and assemble DDB put requests
-  const keys = {};
-  const formatted = await Promise.all(records.map(r => exports.formatWrite(r, idField)));
-  formatted.forEach(fmt => keys[fmt.PutRequest.Item.id.S] = [process.env.DDB_TABLE, fmt]);
+  const params = {
+    AttributeUpdates: {},
+    Key: { id: { S: id } },
+    ReturnValues: 'ALL_OLD',
```

Comment:
This is the real magic.  It returns the state of the DDB record _before_ we updated it.  So we can check which of the `segments` we've already counted in a single transaction.